### PR TITLE
Use `GetMetaData` for stride computation

### PR DESCRIPTION
**Pull request by zasdfgbnm**
The definition of `Tensor` in runtime has been changed as

```CUDA
template <typename T, int Dims, int AllocDims = Dims>
struct Tensor {
  __device__ T& operator[](nvfuser_index_t ind) {
    return data[ind];
  };

  T* data;
  Array<nvfuser_index_t, Dims, 1> logical_size;
  Array<nvfuser_index_t, AllocDims, 1> alloc_stride;
};
```

That is, we are more explicit about whether we are referring to the size/stride of allocation domain or rFactor domain. On the host, the `PolymorphicValue` of tensor metadata now stores all the four `logical_size`, `logical_stride`, `alloc_size`, `alloc_stride`.

The utility that converts sizes and strides wrt rFactor to allocation domain was `inferAndValidateAllocationSizesAndStrides`, this function has been moved to `tensor_metadata.cpp` and is now private to that file. The new way to compute sizes and strides of allocation domain is:

```C++
expression_evaluator.evaluate(IrBuilder::metadataExpr(tv));
```

I need to change `SchedulerRuntimeInfo::SchedulerRuntimeInfo`, `getKernelArgument`, and `validateAlignedVectorizedFusionInputOutput` to use this new approach.

With this change, `getTensorArg` and `KernelArgumentHolder::getBuffer` are no longer used. We should be ready to remove all the subclasses of `ArgAbstract`, but I am not doing this clean up. These cleanups will be left as next PR.

### CODE REVIEW

**Comment by zasdfgbnm** [comment]
Not working

File: csrc/executor_kernel_arg.cpp [hunk 1]
```diff
@@ -14,303 +14,6 @@
 
 namespace nvfuser {
 
-namespace {
-
-// Forward traverse from rFactor domain to allocation domain, compute frontier
-// sizes and strides, validate that splits are divisible and merges are
-// contiguous, and update active_ids_ correspondingly.
-class ForwardTraverseFromRFactorToAlloc {
```
Comment (zasdfgbnm) [inline csrc/executor_kernel_arg.cpp:22]: Moved to `tensor_metadata.cpp` unchanged.

File: csrc/executor_kernel_arg.cpp [hunk 2]
```diff
@@ -14,303 +14,6 @@
...
-        {rfactor.begin(), rfactor.end()},
-        {alloc.begin(), alloc.end()});
-    for (auto expr : forward_exprs) {
-      handle(expr);
-    }
-  }
-};
-
-// Similar to ForwardTraverseFromRFactorToAlloc, but in the opposite direction.
-class BackwardTraverseFromRFactorToAlloc {
```
Comment (zasdfgbnm) [inline csrc/executor_kernel_arg.cpp:117]: Moved to `tensor_metadata.cpp` unchanged.

File: csrc/executor_kernel_arg.cpp [hunk 3]
```diff
@@ -14,303 +14,6 @@
...
-// of its corresponding TensorView, compute the sizes and strides of the tensor
-// with respect to its allocation domain.
-// For example, if the rFactor domain is [I1, I2], and the allocation domain is
-// [I2*I1], and the tensor's size is [5, 3] and stride is [2, 10], then the
-// resulting size will be [15] and stride will be [2]
-// Another example, if the rFactor domain is [I1*I2] and the allocation domain
-// is [I1, I2], and the tensor's size is [15] and stride is [7], and the extent
-// of I2 is 5, then the resulting size will be [3, 5] and stride will be [35, 7]
-std::vector<std::pair<int64_t, int64_t>>
-inferAndValidateAllocationSizesAndStrides(
```
Comment (zasdfgbnm) [inline csrc/executor_kernel_arg.cpp:225]: Moved to `tensor_metadata.cpp` slightly changed.

**Comment by zasdfgbnm** [comment]
!build

**Comment by zasdfgbnm** [comment]
!build

File: csrc/executor_kernel_arg.h [hunk 4]
```diff
@@ -297,22 +290,6 @@ struct TensorArg : public TensorArgAbstract {
...
-  void inferSetAndValidateStrides(
-      const at::Tensor& tensor,
-      TensorView* tv,
-      ExpressionEvaluator& eval) {
-    auto sizes_strides =
-        inferAndValidateAllocationSizesAndStrides(tensor, tv, eval);
-    TORCH_INTERNAL_ASSERT(
-        (size_t)instance_.nAllocationDims() == sizes_strides.size());
-    for (auto i : c10::irange((int64_t)sizes_strides.size())) {
-      alloc_sizes.at(i) = sizes_strides.at(i).first;
```
Comment (jacobhinkle) [inline csrc/executor_kernel_arg.h:312]: Since this is removed, where does `alloc_sizes` get set now?

**Comment by naoyam** [comment]
@zasdfgbnm Could you remind me again what the logical stride means? Does that just strides computed from the logical sizes of a logical contiguous tensor?

**Comment by zasdfgbnm** [comment]
logical strides is just the stride in terms of rFactor domain, that is the "raw" strides from PyTorch using `tensor.strides()`.

**Pull request merged**